### PR TITLE
feat: refactor default component types

### PR DESCRIPTION
**Pull request by mevan-karu**
Related to https://github.com/openchoreo/openchoreo/issues/1951

This pull request focuses on improving robustness and clarity in component handling, sample configurations, and documentation. The main changes include making the code more resilient to missing `ComponentProfile` data, simplifying and updating sample YAML files to use default component types, and refining documentation for easier understanding of API management integration.

**Controller robustness improvements:**
- Added checks in `buildComponentFromRelease` and `applyDefaultNotificationChannel` to safely handle cases where `ComponentProfile` may be nil, preventing potential nil pointer dereferences. [[1]](diffhunk://#diff-440c67a69f0874087394cbd13c783e3bfd5465ecbac9df9f2dce3cf2dbd95fbeR834-R841) [[2]](diffhunk://#diff-440c67a69f0874087394cbd13c783e3bfd5465ecbac9df9f2dce3cf2dbd95fbeL843-R852) [[3]](diffhunk://#diff-440c67a69f0874087394cbd13c783e3bfd5465ecbac9df9f2dce3cf2dbd95fbeR1193-R1197)

**Sample YAML and configuration simplification:**
- Updated `component-http-service` and `component-web-app` sample YAMLs to use the default `deployment/service` and `deployment/web-application` component types, removed custom `ComponentType` definitions, and standardized container names and resource overrides. [[1]](diffhunk://#diff-b8bf5782b36560b81fcf560d06c63a337e8c125186097108451ccbf8404c14d0L1-L98) [[2]](diffhunk://#diff-b8bf5782b36560b81fcf560d06c63a337e8c125186097108451ccbf8404c14d0L109-R22) [[3]](diffhunk://#diff-b8bf5782b36560b81fcf560d06c63a337e8c125186097108451ccbf8404c14d0L130-R40) [[4]](diffhunk://#diff-b8bf5782b36560b81fcf560d06c63a337e8c125186097108451ccbf8404c14d0R60) [[5]](diffhunk://#diff-a03b38594a3ecf2421e331da5ad122ce422331f59061a6c4e2c03ab205640382L1-L87) [[6]](diffhunk://#diff-a03b38594a3ecf2421e331da5ad122ce422331f59061a6c4e2c03ab205640382L99-L102) [[7]](diffhunk://#diff-a03b38594a3ecf2421e331da5ad122ce422331f59061a6c4e2c03ab205640382L120-R38) [[8]](diffhunk://#diff-a03b38594a3ecf2421e331da5ad122ce422331f59061a6c4e2c03ab205640382R60)
- Added the new `worker.yaml` sample to the `make/lint.mk` list for linting.
- Cleaned up `components-with-alert-rules.yaml` by removing redundant `parameters` blocks from component specs. [[1]](diffhunk://#diff-2316747039be627aee22c7c19e4efe82bc66ee5a8972d707b0496b0cc0f207beL15-L25) [[2]](diffhunk://#diff-2316747039be627aee22c7c19e4efe82bc66ee5a8972d707b0496b0cc0f207beL58-L68) [[3]](diffhunk://#diff-2316747039be627aee22c7c19e4efe82bc66ee5a8972d707b0496b0cc0f207beL100-L110)

**Documentation and usability improvements:**
- Updated the `component-with-api-management` README to clarify that API management is now achieved by adding the `api-configuration` trait to the default service component type, rather than using a custom `ComponentType`. The documentation now better explains the flow and configuration, and updates resource names and references for clarity. [[1]](diffhunk://#diff-2b289edb112fd8562a79f0eed5487732c73e3238e897944dfc35c377eace06ffL3-R17) [[2]](diffhunk://#diff-2b289edb112fd8562a79f0eed5487732c73e3238e897944dfc35c377eace06ffL28-R64) [[3]](diffhunk://#diff-2b289edb112fd8562a79f0eed5487732c73e3238e897944dfc35c377eace06ffL99-R75) [[4]](diffhunk://#diff-2b289edb112fd8562a79f0eed5487732c73e3238e897944dfc35c377eace06ffL109-L124)
- Fixed the sample HTTP service README to use the correct endpoint path in `curl` commands, reflecting the new service exposure approach. [[1]](diffhunk://#diff-f561adb08e8dc76f6993121a6ffc5bedb8594752f74883353bdc9813810f9054R76-R79) [[2]](diffhunk://#diff-f561adb08e8dc76f6993121a6ffc5bedb8594752f74883353bdc9813810f9054L86-R88)

### Mini breakdown — changed files by top-level folder
- internal: 1 file changed (internal/controller/releasebinding/controller.go) — +15 / -2
- make: 1 file changed (make/lint.mk) — +1 / 0
- samples: ~38 files changed across samples/ (component-types, from-image, from-source, gcp-microservices-demo, getting-started, etc.) — aggregate approx +632 / -1,224 (net ~-592 lines)
- api/, config/, pkg/, install/, docs/, openapi/, rca-agent/, cmd/: no top-level code changes reported in the diff summary

Total referenced files changed: ~40 files | approx +632 / -1,224 (net ~-592 lines)

### Key changes (evidence-based)
- Controller robustness
  - internal/controller/releasebinding/controller.go: added nil-safety guards in buildComponentFromRelease and applyDefaultNotificationChannel to avoid nil pointer dereferences when ComponentProfile is nil. No public API/signature changes.
- Samples and configuration simplification
  - Standardized default component types in samples:
    - deployment/http-service → deployment/service
    - deployment/web-service → deployment/web-application
    - Replaced custom http-service-with-api-management ComponentType with deployment/service + api-configuration trait.
  - Added worker ComponentType and included worker.yaml in make/lint.mk.
  - Removed many inline Component spec.parameters blocks (replicas, exposed, resources.requests/limits removed) across many sample Components.
  - Moved environment-driven overrides to envOverrides and ReleaseBinding.componentTypeEnvOverrides (many samples add replicas: 2 in release bindings).
  - Standardized container name from app → main across ComponentType templates, Workloads, and trait params.
  - Sample trait updates: introduced api-configuration in samples for API management; HPA parameter renames and wiring updates (targetCPUPercent → targetCPUUtilizationPercentage) in samples; persistent-volume and emptydir-volume trait examples updated.
- Documentation and usability
  - README updates to document base path exposure at /{component-name}, reflect api-configuration trait usage, and correct curl examples.
  - Deleted obsolete README for component-with-configs.

### API / CRD surface changes and compatibility risk
- Controller/CRD runtime code:
  - Only internal nil-safety logic added in controller code; no exported controller API/CRD type changes. Compatibility risk: Low.
- Sample-level public schema/CR changes (affecting consumers who copy samples):
  - New/updated ComponentType samples (worker, service, web-application, scheduled-task) with changed parameter schemas (parameters simplified/removed; envOverrides introduced), containerName removed, container fixed to "main".
  - Trait example changes in samples (new api-configuration trait usage; HPA parameter renames; trait parameter default/name changes).
  - Compatibility risk for consumers copying/updating manifests: Medium — sample field/parameter names and structure changed (replicas/exposed → envOverrides; containerName removal; container renamed to "main"; trait param renames).
- Overall compatibility risk summary:
  - Controller runtime API risk: Low.
  - Sample/consumer compatibility risk: Medium.

### Tests added/updated
- No test files added or modified in this PR (0 tests changed according to provided summary).
- Recommended test coverage gaps:
  - Unit tests for controller nil-safety branches: buildComponentFromRelease and applyDefaultNotificationChannel (nil ComponentProfile paths).
  - Integration/E2E tests: rendering/templating of default component types (service/web-application/worker), ReleaseBinding.componentTypeEnvOverrides application (replica overrides), and api-configuration trait end-to-end effects (Backend/RestApi/HTTPRoute patching).
  - Trait/unit tests: HPA parameter rename and wiring, persistent-volume/emptydir-volume patches.

### Risk hotspots (short reasons)
- Controller nil-safety (Medium)
  - Reason: new early nil-checks change execution paths when ComponentProfile is absent; requires unit tests to ensure behavior remains correct and not silently skipping intended overrides.
- Sample-driven consumption/migration (Medium)
  - Reason: widespread sample signature and parameter changes (replicas/exposed → envOverrides, removal of containerName, default container name change to "main") may break users who reuse existing sample manifests.
- Trait parameter/behavior changes in samples (Medium)
  - Reason: sample trait parameter renames and rewiring (HPA target rename, api-configuration adoption) require updates to any tooling or scripts that rely on prior sample shapes.
- Silent profile skipping (Low–Medium)
  - Reason: nil guards could cause ComponentProfile-related parameters/traits to be ignored when absent; consider adding logging/metrics when guard paths are taken.
- Authn/Authz, RBAC, Secrets (Low)
  - Reason: only sample manifests/docs touched secret references; no controller RBAC/auth code altered.
- Install/Upgrade paths (Low)
  - Reason: no CRD/operator install artifacts or Helm manifests changed — controller and CRDs remain compatible.

### Recommendations
- Add unit tests for controller nil-safety branches (buildComponentFromRelease, applyDefaultNotificationChannel).
- Add integration/E2E tests validating:
  - sample rendering for default component types (service/web-application/worker),
  - ReleaseBinding/componentTypeEnvOverrides behavior (replicas overrides),
  - api-configuration trait behavior in the sample flow.
- Add migration notes in docs for sample consumers describing:
  - parameters renames (replicas/exposed → envOverrides),
  - containerName removal and container rename to "main",
  - trait parameter changes (HPA, api-configuration).

### CODE REVIEW

File: internal/controller/releasebinding/controller.go [hunk 1]
```diff
@@ -1182,6 +1190,11 @@ func (r *Reconciler) applyDefaultNotificationChannel(
 	rb *openchoreov1alpha1.ReleaseBinding,
 	componentRelease *openchoreov1alpha1.ComponentRelease,
 ) error {
+	// Check if ComponentProfile exists
+	if componentRelease.Spec.ComponentProfile == nil {
```
Comment (ChathurangaKCD) [inline internal/controller/releasebinding/controller.go:1194]: is this due to code rabbit comment? 🤔

File: samples/getting-started/component-types/webapp.yaml [hunk 2]
```diff
@@ -14,6 +14,10 @@ spec:
     - api-configuration
     - observability-alert-rule
 
+  validations:
+    - rule: "${workload.endpoints.exists(name, endpoint, endpoint.type == 'HTTP' || endpoint.type == 'REST')}"
```
Comment (ChathurangaKCD) [inline samples/getting-started/component-types/webapp.yaml]:
```suggestion
    - rule: "${workload.endpoints.exists(name, endpoint, endpoint.type == 'HTTP')}"
```

HTTP endpoint is a must, right? Otherwise wouldn't able to serve any html?

File: samples/getting-started/component-types/service.yaml [hunk 3]
```diff
@@ -25,13 +29,12 @@ spec:
         memory: "string | default=256Mi"
 
     parameters:
-      replicas: "integer | default=1"
-      imagePullPolicy: "string | default=IfNotPresent"
       exposed: "boolean | default=false"
```
Comment (ChathurangaKCD) [inline samples/getting-started/component-types/service.yaml]: I thought we made this true, about a week ago. Not sure whether it was set to false intentionally, or because both getting-started samples and that change was done around the same time.

File: internal/controller/releasebinding/controller.go (hunk 1 above)
Comment (mevan-karu) [reply internal/controller/releasebinding/controller.go:1194]: No there's a nil pointer when componenttype doesn't have any params defined.

File: samples/getting-started/component-types/webapp.yaml (hunk 2 above)
Comment (mevan-karu) [reply samples/getting-started/component-types/webapp.yaml]: +1. Fixed.

File: samples/getting-started/component-types/service.yaml (hunk 3 above)
Comment (mevan-karu) [reply samples/getting-started/component-types/service.yaml]: `exposed` param?

Comment (ChathurangaKCD) [reply samples/getting-started/component-types/service.yaml]: https://github.com/openchoreo/openchoreo/pull/1702/changes

File: internal/controller/releasebinding/controller.go
Comment (ChathurangaKCD) [inline internal/controller/releasebinding/controller.go:1224]: Unrelated to this PR. @akila-i Can't we make default notification channel available in cel context and use it? We've hardcoded a trait name here.

File: samples/getting-started/component-types/webapp.yaml [hunk 4]
```diff
@@ -14,6 +14,10 @@ spec:
     - api-configuration
     - observability-alert-rule
 
+  validations:
+    - rule: "${workload.endpoints.exists(name, endpoint, endpoint.type == 'HTTP')}"
+      message: "Web-application components must have at least one HTTP."
```
[the remark this answers is not in the document]
Comment (mevan-karu) [reply samples/getting-started/component-types/webapp.yaml:19]: Fixed

File: samples/getting-started/component-types/service.yaml (hunk 3 above)
Comment (mevan-karu) [reply samples/getting-started/component-types/service.yaml]: Fixed

File: internal/controller/releasebinding/controller.go
Comment (akila-i) [reply internal/controller/releasebinding/controller.go:1224]: Ack. Created https://github.com/openchoreo/openchoreo/issues/1991 to track this.
